Eugen

theugen opened the pull request:
Ho cominciato a lavorare sulla issue #31,  ovvero il ciclo su tutte le frequenze di taglio.
Questa dovrebbe essere una versione grezza (per i dettagli, leggere i commit) ma funzionante,
se nessuno ha fatto tentativi e prodotto versioni migliori propongo di usarla come base per il lavoro successivo.

theugen committed:
Issue #31: primi passi

Ho impostato il ciclo da fare sulle fft da fare a diverse frequenze
di taglio, il lavoro è ancora grezzo e lungi dall'essere finito,
però la struttura generale potrebbe esserci.
L'idea è questa: io sono un ricercatore, ho uno script scritto da terzi,
devo decidere quando tagliare sulla frequenza. A questo punto creo un
file di testo (io ho messo in /data insieme ai dati convertiti, poi ne
possiamo discutere) con i tempi e lui va a leggermelo, è più comodo che
non aprire lo script e andare a modificare il vettore di frequenze nel
sorgente.
Siccome a me quei dati mancano, ho messo una decina di tempi di prova,
poi mettiamo i veri appena li abbiamo.
I tempi vengono poi convertiti in frequenze, e lì comincia il ciclo.
Fuori dal ciclo viene creata una cartella all'interno di quella dei grafici
del singolo dataset in cui vengono sparati tutti i grafici della fft a diverse
frequenze di taglio, così non si mischiano agli altri (saranno tantini)
e rimane più facile cercare quello che si vuole.

Problemi aperti:
- al momento il file viene letto n-mila volte (è nel ciclo delle stazioni):
può valer la pena spostare la lettura nel main fuori dal ciclo, in modo da
risparmiare quel decimo di secondo (che sommato ad altri decimi di secondo
magari fa un bel totale)? Perchè a rigor di logica nel main uno potrebbe anche
decidere di non eseguire lo script periodigramma (e a quel punto il file è inutile
importarlo), oppure di passare frequenze diverse per più stazioni (anche se
non so a che possa servire).
- al momento abbiamo un filtro passa-basso, noi non volevamo forse anche un passa-banda?
- temo che i tempi comincino a diventare lunghetti, bisogna rendere efficiente
al max dove possibile, ed ero quasi tentato dal cercare una soluzione di parallelizzazione
(però di questo ne parleremo nella milestone "Versione 2" :P)...
- i grafici sono proprio bruttini, neanche un titolo :D

 Changes to be committed:
	modified:   Periodigram.R
	modified:   turbulence_class.R

theugen committed:
Aggiunto file cut_freq

cut_freq è il file che ho creato con i tempi (da trasformare in frequenze) di prova. Va spostato in data/.

AnnaN92 commented:
Eugenio, per te è un problema mandare dei grafici di esempio? Solo per comprendere di più. Grazie mille.

theugen commented:
Fatto! Hai sulla mail ;)

theugen committed:
Individuare frequenza meandering

Stefano l'altro giorno ha proposto un metodo per individuare la frequenza
di meandering nella fft: usare un metodo di derivazione numerico per individuare
la pendenza della retta tangente localmente, nella fattispecie a noi interesserebbero
gli estremanti. Lui proponeva di usare diff(), a me è venuto in mente che
avevo già scritto un'interpolazione lineare su finestra mobile che restituisce
il valore del coefficiente angolare calcolato sulla finestra considerata.
Sto facendo le prime prove con valori diversi di finestre, se sono troppo
grandi il picco non viene preso (l'interpolazione rimane negativa), più
diventa piccola più i tempi di calcolo aumentano (la finestra viene spostata
di 1 elemento alla volta): per ora sembra che qualcosa si veda già da una
dimensione pari a 501 elementi, rimane da capire bene se è sensata la cosa ma
sono abbastanza fiducioso.
Se qualcuno vuole provare il pacchetto che ho scritto con le funzioni si trova
https://github.com/theugen/mwindow
basta scaricare, decomprimere e installare. La documentazione è già nel pacchetto
ed è accessibile con le solite modalità.
Vi tengo aggiornati.

theugen committed:
Dettagli minori

Aggiunto qualche dettaglio minore al commit precedente, leggere quello.

theugen committed:
Merge branch 'Eugen' of https://github.com/Atmosferica/Turbolenza into Eugen

theugen committed:
Sistemato grafico finestra

Ho sistemato un po' il grafico della finestra mobile,
rimangono da mettere giù bene gli assi (trovare estremi,
arrotondare, titoli) e trovare la dimensione corretta della
finestra. Sovrapporre al grafico fft?!?

theugen committed:
Velocizzato periodigramma + finestra mobile

Io comincio ad avere qualche problema di velocità sul periodigramma,
per cui ho fatto prove per parallelizzare il ciclo for() sulle frequenze
di taglio. Al momento prima fa una chiamata a Sys.info() per determinare
il sistema operativo,  poi se è linux lancia il processo in parallelo
(usando N cores -1, per dar modo di poter usare il computer anche per fare
altro), se è windows lancia il for() normale (bisognerebbe usare una parallelizzazione
diversa, ma non ho modo di provare. Se qualcuno fa da beta tester posso anche
provare a implementare).
In questo modo, usando 3 cores, ci metto circa il 40% del tempo che ci metteva prima.
Allo studio dei benchmark seri.

Riguardo alla questione della finestra mobile, per me non funziona... è troppo poco
sensibile (i picchi sono troppo stretti), ho provato ma se tengo la finestra troppo
larga prende un solo picco nel pieno del rumore bianco (con ampiezza trascurabile),
se la restringo prende troppi picchi. Probabilmente bisognerebbe inserire delle
condizioni sull'ampiezza (considerare solo il picco ad ampiezza maggiore?),
ma non sono sicuro che funzioni.
Provare la strada del diff()?

algebrato commented:
Per quanto riguarda i tempi di esecuzione il più del tempo mi sa che viene usato per fare questo:
`for(...){`
`data <- dofft(vel, 10)`
`filt <- filter.data(data$freq, data$fft_vel, 10)#, cut_freq[1,1])`
`}`

in ogni ciclo `data` è sempre lo stesso che viene ricalcolato ogni volta. In teoria andrebbe fuori dal ciclo. Questo dovrebbe aiutare un po' tanto il miglioramento dei tempi di esecuzione.
Fai conto che la la FFT inversa, ad ogni step del ciclo, diventa sempre più veloce da calcolare.
Se tagli ad 1Hz, la fft inversa la fai considerendo informazioni fino appunto ad 1Hz, se tagli a 0.01Hz la FFT la fai solo coi dati da 0 a 0.01Hz. Mentre ora mi sa che viene preso un po' tutto (compresi gli 0+0i).

Prima di iniziare a parallelizzare proverei prima a semplificare le varie parti di codice ridondante. Apro una issue su queste cose a cui mi assegno. (Però riuscirò a lavorarci solo da domani ... oggi non riesco)

algebrato commented:
Add some fix and feautures by Eugen branch.

Merge del branch di Eugen. Ci sono vari fix e una proposta di determinazione della frequenza di meandering. Rimangono comunque aperte la issue #28  #35  #36

theugen commented:
Orca, mi è stato approvato tutto, grazie! :smile: 

Ti do ragione su tutti i fronti: a mia discolpa spiego un attimo meglio il perchè della parallelizzazione.
Sono stato "costretto" ad implementarla perchè sto facendo la prova del ciclo sulle frequenze di taglio sul dataset delle 24 ore che hai già usato anche tu: il mio non è un pc particolarmente performante, se voglio avere qualche speranza di finire almeno qualche iterazione del ciclo devo per forza mandare in parallelo (adesso è da due ore che ho 3 core a palla e non ha ancora finito un'iterazione). Poi va beh, è innegabile che mi sia divertito!

Sui dataset più piccoli su cui ho svolto le prove invece non ci sono particolari problemi, si passa da 4 minuti a poco meno di 2, in quel caso è molto più importante l'ottimizzazione del codice, su cui continuerei comunque a puntare tutto, quindi grazie mille per aver creato la nuova issue perchè effettivamente è importante ;)
(Piccola nota: con la finestra mobile, su singolo core, si aggiungevano altri 5 minuti di calcolo circa, al momento l'ho commentata perchè non mi sembra si quagli molto...)
(Piccola nota 2: son contento che il pacchetto funzioni)
